primo/opt_model/model_options.py

```diff
@@ -0,0 +1,350 @@
+
+        index : list
+            List of rows/indices in the WellData object
+
+        column : str
+            Allowed columns are "age", "depth"
+        """
+        wd = self.config.well_data
+        df = wd.data
+        column = getattr(wd._col_names, column)
+        return {
+            (j, k): abs(df.loc[j, column] - df.loc[k, column])
+            for j, k in combinations(index, 2)
+        }
+
+    def build_optimization_model(self):
+        """Builds the optimization model"""
+        LOGGER.info("Beginning to construct the optimization model.")
+        self._opt_model = PluggingCampaignModel(self)
+        return self._opt_model
+
+    def solve_model(self, **kwargs):
+        """Solves the optimization"""
+
+        # Adding support for pool search if gurobi_persistent is available
+        # To get n-best solutions, pass pool_search_mode = 2 and pool_size = n
+        pool_search_mode = kwargs.pop("pool_search_mode", 0)
+        pool_size = kwargs.pop("pool_size", 10)
+
+        # If a solver is specified, use it.
+        if "solver" in kwargs:
+            solver = get_solver(**kwargs)
+
+        else:
+            # Otherwise, auto-detect solver in order of priority
+            for sol in ("gurobi_persistent", "gurobi", "scip", "glpk", "highs"):
+                if SolverFactory(sol).available(exception_flag=False):
+                    LOGGER.info(
+                        f"Optimization solver is not specified. "
+                        f"Using {sol} as the optimization solver."
+                    )
+                    solver = get_solver(solver=sol, **kwargs)
+                    break
+
+        self._solver = solver
+        if solver.name == "gurobi_persistent":
+            # For persistent solvers, model instance need to be set manually
+            solver.set_instance(self._opt_model)
+            solver.set_gurobi_param("PoolSearchMode", pool_search_mode)
+            solver.set_gurobi_param("PoolSolutions", pool_size)
+
+        # Solve the optimization problem
+        solver.solve(self._opt_model)
+
+        return self._opt_model.get_optimal_campaign()
+
+    def get_solution_pool(self):
+        """Extracts solutions from the solution pool"""
+        solver = self._solver
+        pm = self._opt_model  # This is the pyomo model
+        gm = solver._solver_model  # This is the gurobipy model
+        # Get Pyomo var to Gurobipy var map.
+        # Gurobi vars can be accessed as py_to_gp[<pyomo var>]
+        pm_to_gm = solver._pyomo_var_to_solver_var_map
+
+        # Return if the pool search mode is not 2
+        if gm.Params.PoolSearchMode != 2 or solver.name != "gurobi_persistent":
+            LOGGER.warning("Pool-search was not used for the solve.")
+            return None
+
+        # Number of solutions found
+        num_solutions = gm.SolCount
+        solution_pool = {}
+        wd = self.config.well_data  # Well data
+
+        for i in range(num_solutions):
```

Comment:
I am also with Will here where even for iterators or function arguments, I prefer "meaningful" variable names. "for cluster in pm.set_clusters" is preferable to "for c in pm.set_clusters" but am conscious this is nitpicky